gonat.go

```diff
@@ -0,0 +1,177 @@
+package gonat
+
+import (
+	"net"
+	"time"
+
+	"github.com/getlantern/errors"
+	"github.com/getlantern/golog"
+	"github.com/oxtoacart/bpool"
+)
+
+const (
+	// DefaultBufferPoolSize is 10 MB
+	DefaultBufferPoolSize = 10000000
+
+	// DefaultBufferDepth is 250 packets
+	DefaultBufferDepth = 250
+
+	// DefaultIdleTimeout is 65 seconds
+	DefaultIdleTimeout = 65 * time.Second
+
+	// DefaultStatsInterval is 15 seconds
+	DefaultStatsInterval = 15 * time.Second
+
+	// MinConntrackTimeout sets a lower bound on how long we'll let conntrack entries persist
+	MinConntrackTimeout = 1 * time.Minute
+
+	// MaximumIPPacketSize is 65535 bytes
+	MaximumIPPacketSize = 65535
+)
+
+const (
+	minEphemeralPort  = 32768
+	maxEphemeralPort  = 61000 // consistent with most Linux kernels
+	numEphemeralPorts = maxEphemeralPort - minEphemeralPort
+)
+
+var (
+	log = golog.LoggerFor("gonat")
+)
+
+type Server interface {
+	// Serve starts processing packets and blocks until finished
+	Serve() error
+
+	// Count of accepted packets
+	AcceptedPackets() int
+
+	// Count of rejected packets
+	RejectedPackets() int
+
+	// Number of TCP connections being tracked
+	NumTCPConns() int
+
+	// Number of UDP connections being tracked
+	NumUDPConns() int
+
+	// Close stops the server and cleans up resources
+	Close() error
+}
+
+type Opts struct {
+	// IFName is the name of the interface to use for connecting upstream.
+	// If not specified, this will use the default interface for reaching the
+	// Internet.
+	IFName string
+
+	// IFAddr is the address to use for outbound packets. Overrides the IFName
+	// when specified.
+	IFAddr string
+
+	// BufferPool is a pool for buffers. If not provided, default to a 10MB pool.
+	// Each []byte in the buffer pool should be 65536 bytes.
+	BufferPool BufferPool
+
+	// BufferDepth specifies the number of outbound packets to buffer between
+	// stages in the send/receive pipeline. The default is 250.
+	BufferDepth int
+
+	// IdleTimeout specifies the amount of time before idle connections are
+	// automatically closed. The default is 65 seconds.
+	IdleTimeout time.Duration
+
+	// StatsInterval controls how frequently to display stats. Defaults to 15
+	// seconds.
```

Comment:
Since the corresponding constants are exported, we could change these comments to refer to the constants (e.g. `Defaults to DefaultStatsInterval`.  That way the comments stay up-to-date if we decide to play around with the constants.

Reply:
Good idea, doing it.